File: src/leetcodepython/array/longest_consecutive_sequence _128.py
# encoding='utf-8'

'''
/**
 * This is the solution of No. 128 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/longest-consecutive-sequence
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个未排序的整数数组，找出最长连续序列的长度。
 *
 * 要求算法的时间复杂度为 O(n)。
 *
 * 示例:
 *
 * 输入: [100, 4, 200, 1, 3, 2]
 * 输出: 4
 * 解释: 最长连续序列是 [1, 2, 3, 4]。它的长度为 4。
 *
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/longest-consecutive-sequence
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''
from typing import List


class Solution:
    def longest_consecutive(self, nums: List[int]) -> List[int]:
        '''
            找出最大连续子串
        Args:
            nums: 数组

        Returns:
            连续子串长度
        '''
        if len(nums) <= 1:
            return len(nums)
        nums.sort()
        max_length, count = 1, 1
        for i in range(len(nums) - 1):
            if nums[i] + 1 == arr[i + 1]:
                count += 1
            else:
                if nums[i] != nums[i + 1]:
                    count = 1
            max_length = max(count, max_length)
        return max_length

    def longest_consecutive_2(nums: List[int]) -> int:
        '''
            找出最大的连续子串
        Args:
            nums: 数组
        Returns:
            最大子串长度
        '''
        if len(nums) <= 1:
            return len(nums)
        count = 0
        num_set = set(nums)
        for num in num_set:
            if num-1 not in num_set:
                current_count = 1
                current_num = num
                while current_num + 1 in nums:
                    current_count += 1
                    current_num += 1
                count = max(count, current_count)
        return count


if __name__ == '__main__':
    arr = [2, 1, 3, 3, 4]
    solution = Solution()
    max_length = solution.longest_consecutive(arr)
    print(max_length)
    assert max_length == 4

File: src/leetcodejava/array/LongestConsecutiveSequence128.java
package leetcodejava.array;

import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * This is the solution of No. 128 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/longest-consecutive-sequence
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个未排序的整数数组，找出最长连续序列的长度。
 * <p>
 * 要求算法的时间复杂度为 O(n)。
 * <p>
 * 示例:
 * <p>
 * 输入: [100, 4, 200, 1, 3, 2]
 * 输出: 4
 * 解释: 最长连续序列是 [1, 2, 3, 4]。它的长度为 4。
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/longest-consecutive-sequence
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class LongestConsecutiveSequence128 {
    @Test
    public void longestConsecutiveSequenceTest() {
        int[] nums = {12, 13, 1, 4, 5, 6};
        int result = longestConsecutive(nums);
        System.out.println(result);
        Assert.assertEquals(result, 3);
    }

    /**
     * 连续数的长度
     *
     * @param nums 数组
     * @return 连续数组长度
     */
    private int longestConsecutive(int[] nums) {
        if (nums == null || nums.length <= 1) {
            return nums.length;
        }
        Arrays.sort(nums);
        int maxLength = 1;
        int count = 1;
        for (int i = 0; i < nums.length - 1; i++) {
            if (nums[i] + 1 == nums[i + 1]) {
                count++;
            } else {
                if (nums[i] != nums[i + 1]) {
                    count = 1;
                }
            }
            maxLength = Math.max(count, maxLength);
        }
        return maxLength;
    }

    /**
     * 最长连续传
     *
     * @param nums 数组
     * @return 最大长度
     */
    public int longestConsecutive2(int[] nums) {
        if (nums == null || nums.length < 1) {
            return 0;
        }
        Set<Integer> numSet = new HashSet();
        for (int num : nums) {
            numSet.add(num);
        }
        int longestStreak = 0;
        for (int num : numSet) {
            if (!numSet.contains(num - 1)) {
                int currentNum = num;
                int currentStreak = 1;
                while (numSet.contains(currentNum + 1)) {
                    currentNum += 1;
                    currentStreak += 1;
                }
                longestStreak = Math.max(longestStreak, currentStreak);
            }
        }
        return longestStreak;
    }
}
